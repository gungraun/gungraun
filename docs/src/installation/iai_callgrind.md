# Gungraun

Gungraun is divided into the library `iai-callgrind` and the benchmark runner
`iai-callgrind-runner`.

## Installation of the library

To start with Gungraun, add the following to your `Cargo.toml` file:

```toml
[dev-dependencies]
iai-callgrind = "0.16.1"
```

or run

```bash
cargo add --dev iai-callgrind@0.16.1
```

## Installation of the benchmark runner

To be able to run the benchmarks you'll also need the `iai-callgrind-runner`
binary installed somewhere in your `$PATH`. Otherwise, there is no need to
interact with `iai-callgrind-runner` as it is just an implementation detail.

### From Source

```shell
cargo install --version 0.16.1 iai-callgrind-runner
```

There's also the possibility to install the binary somewhere else and point the
`IAI_CALLGRIND_RUNNER` environment variable to the absolute path of the
`iai-callgrind-runner` binary like so:

```shell
cargo install --version 0.16.1 --root /tmp iai-callgrind-runner
IAI_CALLGRIND_RUNNER=/tmp/bin/iai-callgrind-runner cargo bench --bench my-bench
```

### Binstall

The `iai-callgrind-runner` binary is
[pre-built](https://github.com/gungraun/gungraun/releases/tag/v0.16.1)
for most platforms supported by valgrind and easily installable with
[binstall](https://github.com/cargo-bins/cargo-binstall)

```shell
cargo binstall iai-callgrind-runner@0.16.1
```

## Updating

When updating the `iai-callgrind` library, you'll also need to update
`iai-callgrind-runner` and vice-versa or else the benchmark runner will exit
with an error.

### In the Github CI

Since the `iai-callgrind-runner` version must match the `iai-callgrind` library
version it's best to automate this step in the CI. A job step in the github
actions CI could look like this

```yaml
- name: Install iai-callgrind-runner
  run: |
    version=$(cargo metadata --format-version=1 |\
      jq '.packages[] | select(.name == "iai-callgrind").version' |\
      tr -d '"'
    )
    cargo install iai-callgrind-runner --version $version
```

Or, speed up the overall installation time with `binstall` using the
[taiki-e/install-action](https://github.com/taiki-e/install-action)

```yaml
- uses: taiki-e/install-action@cargo-binstall
- name: Install iai-callgrind-runner
  run: |
    version=$(cargo metadata --format-version=1 |\
      jq '.packages[] | select(.name == "iai-callgrind").version' |\
      tr -d '"'
    )
    cargo binstall --no-confirm iai-callgrind-runner --version $version
```
