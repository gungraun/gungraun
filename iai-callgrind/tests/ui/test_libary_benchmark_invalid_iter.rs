use iai_callgrind::library_benchmark;

mod test_when_together_with_other_main_parameter {
    use super::*;

    #[library_benchmark]
    #[benches::some(iter = 1..=2, args = [0, 1])]
    fn bench_10() -> u64 {
        10
    }
}

mod test_when_iter_twice {
    use super::*;

    #[library_benchmark]
    #[benches::some(iter = 1..=2, iter = vec![0, 1])]
    fn bench_10(a: u64) -> u64 {
        a + 1
    }
}

mod test_when_iter_no_bench_arg {
    use super::*;

    #[library_benchmark]
    #[benches::some(iter = 1..=2)]
    fn bench_10() -> u64 {
        1
    }
}

mod test_when_iter_too_many_bench_arg {
    use super::*;

    #[library_benchmark]
    #[benches::some(iter = 1..=2)]
    fn bench_10(a: u64, b: u64) -> u64 {
        a + b
    }
}

mod test_when_wrong_argument_type {
    use super::*;

    #[library_benchmark]
    #[benches::some(iter = 1..=2)]
    fn bench_10(a: &str) -> String {
        a.to_owned()
    }
}

mod test_when_setup_then_wrong_argument_type {
    use super::*;

    fn setup(a: u64) -> String {
        a.to_string()
    }

    #[library_benchmark]
    #[benches::some(iter = 1..=2, setup = setup)]
    fn bench_10(a: u64) -> String {
        a.to_string()
    }
}

mod test_when_setup_has_wrong_argument_type {
    use super::*;

    fn setup(a: String) -> u64 {
        a.parse().unwrap()
    }

    #[library_benchmark]
    #[benches::some(iter = 1..=2, setup = setup)]
    fn bench_10(a: String) -> String {
        format!("foo: {a}")
    }
}

mod test_when_teardown_has_wrong_argument_type {
    use super::*;

    fn teardown(a: u64) {
        println!("{a}");
    }

    #[library_benchmark]
    #[benches::some(iter = 1..=2, teardown = teardown)]
    fn bench_10(a: u64) -> String {
        format!("foo: {a}")
    }
}

mod test_when_not_an_iterator {
    use super::*;

    #[library_benchmark]
    #[benches::some(iter = 1)]
    fn bench_10(a: u64) -> String {
        format!("foo: {a}")
    }
}

fn main() {}
