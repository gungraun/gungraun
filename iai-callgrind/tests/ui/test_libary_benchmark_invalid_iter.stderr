error: Only one parameter of `file`, `args` or `iter` can be present
 --> tests/ui/test_libary_benchmark_invalid_iter.rs:7:16
  |
7 |     #[benches::some(iter = 1..=2, args = [0, 1])]
  |                ^^^^

error: Duplicate argument: `iter`

         = help: `iter` is allowed only once

  --> tests/ui/test_libary_benchmark_invalid_iter.rs:17:35
   |
17 |     #[benches::some(iter = 1..=2, iter = vec![0, 1])]
   |                                   ^^^^^^^^^^^^^^^^^

error: The benchmark function can only take exactly one argument if the iter parameter is present

         = help: fn benchmark_function(arg: String) ...
         = note: If you need more than one argument you can use a tuple as input and
           destruct it in the function signature. Example:

           #[benches::some_id(iter = vec![(1, 2)])]
           fn benchmark_function((first, second): (u64, u64)) -> usize { ... }

  --> tests/ui/test_libary_benchmark_invalid_iter.rs:28:8
   |
28 |     fn bench_10() -> u64 {
   |        ^^^^^^^^

error: The benchmark function can only take exactly one argument if the iter parameter is present

         = help: fn benchmark_function(arg: String) ...
         = note: If you need more than one argument you can use a tuple as input and
           destruct it in the function signature. Example:

           #[benches::some_id(iter = vec![(1, 2)])]
           fn benchmark_function((first, second): (u64, u64)) -> usize { ... }

  --> tests/ui/test_libary_benchmark_invalid_iter.rs:38:8
   |
38 |     fn bench_10(a: u64, b: u64) -> u64 {
   |        ^^^^^^^^

error[E0308]: mismatched types
  --> tests/ui/test_libary_benchmark_invalid_iter.rs:46:5
   |
46 |     #[library_benchmark]
   |     ^^^^^^^^^^^^^^^^^^^^ expected `&str`, found integer
47 |     #[benches::some(iter = 1..=2)]
   |                ---- arguments to this function are incorrect
   |
note: function defined here
  --> tests/ui/test_libary_benchmark_invalid_iter.rs:47:16
   |
47 |     #[benches::some(iter = 1..=2)]
   |                ^^^^
48 |     fn bench_10(a: &str) -> String {
   |                 -------
   = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> tests/ui/test_libary_benchmark_invalid_iter.rs:60:5
   |
60 |     #[library_benchmark]
   |     ^^^^^^^^^^^^^^^^^^^^ expected `u64`, found `String`
61 |     #[benches::some(iter = 1..=2, setup = setup)]
   |                ---- arguments to this function are incorrect
   |
note: function defined here
  --> tests/ui/test_libary_benchmark_invalid_iter.rs:61:16
   |
61 |     #[benches::some(iter = 1..=2, setup = setup)]
   |                ^^^^
62 |     fn bench_10(a: u64) -> String {
   |                 ------
   = note: this error originates in the attribute macro `library_benchmark` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0631]: type mismatch in function arguments
  --> tests/ui/test_libary_benchmark_invalid_iter.rs:75:43
   |
70 |     fn setup(a: String) -> u64 {
   |     -------------------------- found signature defined here
...
75 |     #[benches::some(iter = 1..=2, setup = setup)]
   |                                           ^^^^^ expected due to this
   |
   = note: expected function signature `fn({integer}) -> _`
              found function signature `fn(String) -> _`
note: required by a bound in `Option::<T>::map`
  --> $RUST/core/src/option.rs
   |
   |     pub fn map<U, F>(self, f: F) -> Option<U>
   |            --- required by a bound in this associated function
   |     where
   |         F: FnOnce(T) -> U,
   |            ^^^^^^^^^^^^^^ required by this bound in `Option::<T>::map`

error[E0308]: mismatched types
  --> tests/ui/test_libary_benchmark_invalid_iter.rs:89:46
   |
89 |     #[benches::some(iter = 1..=2, teardown = teardown)]
   |                                              ^^^^^^^^
   |                                              |
   |                                              expected `u64`, found `String`
   |                                              arguments to this function are incorrect
   |
note: function defined here
  --> tests/ui/test_libary_benchmark_invalid_iter.rs:84:8
   |
84 |     fn teardown(a: u64) {
   |        ^^^^^^^^ ------

error[E0689]: can't call method `into_iter` on ambiguous numeric type `{integer}`
  --> tests/ui/test_libary_benchmark_invalid_iter.rs:99:28
   |
99 |     #[benches::some(iter = 1)]
   |                            ^
   |
help: you must specify a type for this binding, like `i32`
   |
98 |     #[library_benchmark]: i32
   |                         +++++
